Add integration with MLflow Tracing

B-Step62 opened the pull request:
## Title

Add integration with MLflow Tracing

## Relevant issues

NA

## Type

🆕 New Feature

## Changes

Hi team! I'm Yuki Watanabe, a maintainer from MLflow. We've released a LLM tracing feature early in this year and we would like to integrate it with LiteLLM. LiteLLM is widely adapted OSS tools so we believe this integration benefit many MLflow users and OSS community🙂

To achieve that, this PR introduces a new logger to create MLflow spans around LLM call. The generated trace looks like this on MLflow UI (and can be exported to any OpenTelemetry collector too, like Jeager, Prometheus, etc).

<img src="https://github.com/user-attachments/assets/d9d356b8-397a-400b-8844-29ec87cc29e7" width="800"/> 


While I validate the code change on my local environment and it works fine, I have two open questions. I would really appreciate if you could provide some guidance for these:

**Q1: Testing** 

It appears that the unit tests for existing handlers are minimal - it does not validate traces are being logged, but seems only validating it does not interrupt the LLM call. I guess this is intentional for not including external tools/service as testing integrations. I'm curious what is the common testing strategy for those integrations - do they run some test suites in their own CI, or perhaps there is separate integration suite for this repository?

**Q2: Synchronous Logging**

Briefly discussed in [another thread](https://github.com/BerriAI/litellm/issues/5216), we are facing one challenge in completing this integration. The problem is that the `log_success_event` handler is executed in a background thread in LiteLLM ([code](https://github.com/BerriAI/litellm/blob/main/litellm/utils.py#L917-L920)), which totally makes sense in general cases but causes a trouble in MLflow Tracing particularly on Databricks notebook.

MLflow tracing has a feature of displaying trace widget inline on Databricks notebook. When users invokes LLM or run chain including it, we generate a trace locally and visualize it in the cell output, so that users can check trace without opening another page and visit trace list. This works like this:

<img src="https://github.com/user-attachments/assets/b6d9a772-f443-47ab-a2ba-f17f5df53fb9"  width="500"/>


However, with LiteLLM, since the success handler is executed in background, the notebook cell often finishes earlier than the trace generation. Then we cannot display traces inline, and even worse, it will be popped up when running another cell, like the following screenshot. This is not acceptable UX and hinder Databricks users from using this LiteLLM.

<img src="https://github.com/user-attachments/assets/c5a0923c-3b1c-4512-ad55-6db945dd3987"  width="500"/>

As a quick and safe solution, I was wondering if we can make the success event logging synchronous when MLflow logger is configured. The code change and impact would be minimal:
```
 if is_mlflow_logger_enabled():
      logging_obj.success_handler(cached_result, start_time, end_time, cache_hit)
 else:
      # current way of calling success handler
      threading.Thread(
          target=logging_obj.success_handler,
          args=(cached_result, start_time, end_time, cache_hit),
      ).start()
```
Alternatively, we can patch the logging utility within MLflow side to make it synchronous.


## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
If UI changes, send a screenshot/GIF of working UI fixes

See the screenshots above. Also curious if there is some way to test tracing integration.

B-Step62 committed:
Add MLflow logger

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Streaming handling

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
lint

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
@krrishdholakia @ishaan-jaff I've implemented MLflow integration for tracing. Would you mind taking a look and hear your thoughts on the two questions in the description? Thank you in advance for your support.

krrishdholakia commented on litellm/integrations/mlflow.py:
```diff
@@ -0,0 +1,241 @@
+from litellm.integrations.custom_logger import CustomLogger
+from litellm.utils import verbose_logger
+
+import mlflow
```
if this is a required import it should be moved inside the class, so it's only required when using mlflow

krrishdholakia commented on litellm/integrations/mlflow.py:
```diff
@@ -0,0 +1,241 @@
+            )
+            span.add_event(event)
+
+            self._end_span_or_trace(
```
is mlflow opentelemetry compatible? this looks pretty similar to OTEL, and i wonder if we can just leverage that integration here

krrishdholakia commented:
Hey @B-Step62 it looks like mlflow is OTEL compatible - https://mlflow.org/docs/latest/llms/tracing/index.html#using-opentelemetry-collector-for-exporting-traces


We already support opentelemetry - https://github.com/BerriAI/litellm/blob/d9e7818e6bcc029ad6197c880b3c4d0386f4f2d6/litellm/integrations/opentelemetry.py#L4

is there a way to leverage that integration instead? (perhaps similar to how arize ai does it - https://github.com/BerriAI/litellm/blob/d9e7818e6bcc029ad6197c880b3c4d0386f4f2d6/litellm/integrations/arize_ai.py#L28

B-Step62 commented:
@krrishdholakia Thank you so much for the review!

> Hey @B-Step62 it looks like mlflow is OTEL compatible - https://mlflow.org/docs/latest/llms/tracing/index.html#using-opentelemetry-collector-for-exporting-traces

Right, MLflow is compatible in terms of the span schema, however, our collector is not compatible with standard OpenTelemetry exporter. In other words:
1. We can export MLflow trace as OTel span.
2. We cannot export arbitral OTel span to MLflow.

IIUC, we need to support 2 for using existing integration, which is not realistic due to multiple backend constraint.

B-Step62 committed:
address comments and fix issues

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
address comments and fix issues

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
@krrishdholakia just following up on my previous comment, let me know if there’s anything I can clarify or if you have additional feedback. Thanks!

krrishdholakia commented on litellm/litellm_core_utils/litellm_logging.py:
```diff
@@ -1248,6 +1250,20 @@ def success_handler(  # noqa: PLR0915
                                 print_verbose=print_verbose,
                             )
+
+                    if callback == "mlflow":
```
this should not be added here. Since the callback is a custom logger, it should work based on line 559

krrishdholakia replied:
make sure the object is instantiated here - https://github.com/BerriAI/litellm/blob/2c37aad1c44af9e98a0af4e36d0b89f2449bf2a9/litellm/litellm_core_utils/litellm_logging.py#L2139

B-Step62 replied:
Makes sense, thank you for the pointer!

B-Step62 replied:
@krrishdholakia When I move the logger instantiation to `_init_custom_logger_compatible_class` and `get_custom_logger_compatible_class`, I noticed that the callback is not called for async APIs like `acomplete`. Is there any way to make the callback invoked for async call as well?

https://github.com/BerriAI/litellm/blob/6b03998e324cd063a3d8133c3388ae21572a740d/litellm/litellm_core_utils/litellm_logging.py#L1183-L1198

krrishdholakia replied:
that sounds wrong. 

the custom logger is called across both, can confirm it works as it's used across otel, prometheus, etc. loggers

B-Step62 replied:
@krrishdholakia But the code indicates the `log_success_event` handler is not invoked for async calls. Am I missing something to make it work?
https://github.com/BerriAI/litellm/blob/6b03998e324cd063a3d8133c3388ae21572a740d/litellm/litellm_core_utils/litellm_logging.py#L1183-L1198

B-Step62 replied:
@krrishdholakia Figured out - the logger needed to implement `async_log_success_event` and confirmed the logging works after adding it.

krrishdholakia commented:
Hey @B-Step62 requested 1 change, 

and if possible can you add a doc on this integration so users know how to enable this. 

See argilla pr for reference - https://github.com/BerriAI/litellm/commit/4f5ff65882754a47f0e12d8dc773dc9904ecb758

B-Step62 committed:
Move logger construction code

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add docs

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
@krrishdholakia I've updated the instantiation part and added documentation. Would you mind giving some thoughts for my question about async calls [here](https://github.com/BerriAI/litellm/pull/6147#discussion_r1823602449)?

B-Step62 committed:
async handlers

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
@krrishdholakia Is this PR ready to merge, or are there any additions needed?

B-Step62 committed:
new picture

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>
